### ISSUE TITLE
Blurry Canvas (image&text) when setDimensions is called

### ISSUE THREAD

**Issue by philippemiguet** [kangax/fabric.js#2791]
Hi all :)

Fabric Version: 1.6.0-rc.1
Fiddle: https://jsfiddle.net/philippemiguet/Lay5t3et
(Hit button and watch the images blur)

What I did:
 - Create canvas 484x484
 - Add image
 - set dimensions 484x484 via `setDimensions` (any dimensions do the same)

What I expected:
 - Image to look at the original with the same quality

What happened:
 - Image was seen to be blurred after `setDimensions` is called

Investigation:
When setDimensions is called on canvas, the image becomes blurry. Using `canvas.imageSmoothingEnabled = false` doesn't solve the problem the image still gets blurry.

However, when the dimensions are set during the initialization the image looks as the original. Here is a [fiddle](https://jsfiddle.net/philippemiguet/Lay5t3et).

I think it is related to the issue #2459 that was fixed here #2337.

**EDIT**: Not only the image get's blurry but the whole canvas. If you add some text it also gets blurry. I have updated the jsfiddle with some text to show you.

**Comment by ozooner** [comment]
I also ran into similar issue. For me the use-case was to initialize canvas based on user's screen size. Since I needed to do it only once, I could use DOM element properties instead:

```
var element = document.createElement('canvas');
element.id = 'cnv';
element.height = window.innerHeight / 2;
element.width = window.innerWidth / 2;
document.body.appendChild(element);
canvas = new fabric.Canvas('cnv');

```
However some library methods still call setDimensions, e.g `toDataURL` with a multiplier parameter

**Comment by philippemiguet** [comment]
Indeed @ozooner good finding, it means generated image will be blurred. Does someone have a hunch what is the issue here?

**Comment by asturur** [comment]
Guys i cannot seen anny blurriness.
Can we check with some image that make  the effect more evidente? form the 4 canvases looks same in couples ( the one with imageSmoothing enabled / not enabled )

**Comment by philippemiguet** [comment]
Hey @asturur you really don't see the text and image getting blurred when clicking on the button? http://cl.ly/253X3H2n1m3a

**Comment by ozooner** [comment]
@asturur 

I actually didn't see the difference in his examples too, but it becomes more apparent when large image is scaled down: https://jsfiddle.net/3m66myvb/2/

**Comment by philippemiguet** [comment]
@asturur @ozooner Sorry guys, maybe I did a bad job as explaining with the fiddle :)

I actually call `setDimensions` when you click the  button, and this is when it is happening. It gets blurry/pixelated. I've made a cleaner jsfiddle: https://jsfiddle.net/philippemiguet/Lay5t3et/8/

I've also made a screencast to show you what I see just in case: http://cl.ly/2Q3x1e283x2r

Are you seeing it this time?

**Comment by philippemiguet** [comment]
I have made a new jsfiddle one with text only: https://jsfiddle.net/philippemiguet/4vw8znkd/7/

As you can see the second text seems blurry and not as sharp as the first one

**Comment by asturur** [comment]
Check my magnified screenshot.

![image](https://cloud.githubusercontent.com/assets/1194048/13174583/fc7dfcf2-d704-11e5-94c2-5a9bbe438c66.png)

Do you see differences?

**Comment by ozooner** [comment]
Interesting...this is how it is for me:
![text](https://i.imgur.com/i51FEW4.png "Text")

**Comment by philippemiguet** [comment]
@asturur @ozooner do you have a retina screen? 

I have a Retina one and I experience the blur on Chrome / Firefox / Safari. 

My friend tried on a non-retina and he can't spot any differences.

**Comment by ozooner** [comment]
Not sure if non-mac screens qualify as retina, but my resolution is 2560x1440

**Comment by asturur** [comment]
if you are a retina user, please download the latest source from here and build it. You should be safe. Otherwise there is still some bug.

**Comment by philippemiguet** [comment]
@asturur indeed it does work with the last build :) Sorry for not testing before with this latest version. Maybe we can add in the Contributing.md to test with the latest build before opening an issue? :) We can close for me.

Also, one tiny question, are the last commit gonna be part of the new release?

**Comment by asturur** [comment]
i'm plan to do a nightly builder.

**Comment by philippemiguet** [comment]
Thanks @asturur for the answer! You rock!

**Comment by maertz** [comment]
I'm a little confused, I'm still struggling with this issue and I don't understand what's wrong.
Since 1.5.0+ something changed which influences the output crispiness. I tried to trace it down but I had no luck. I've tested this on two machines, one with retina display, another without retina display. I've disabled "enableRetinaScaling" to compare both versions with each other.

I understand that if "enableRetinaScaling" has been disabled the visual on the actual screen might be blurry" but when I export the output (with the actual size of the canvas) the results should be the same as on "non retina displays" - right?

## RETINA DISPLAY
**Fabric 1.5.0**
- Chrome Version 54.0.2840.98 (64-bit)
- enabledRetinaScaling: false (not supported in 1.5.0)
- devicePixelRatio: 1 (not supported in 1.5.0)
![fabric 1 5 0 dpr 1](https://cloud.githubusercontent.com/assets/921705/20632243/66cff754-b33c-11e6-8f60-f57c2d12a772.jpeg)
![fabric 1 5 0 dpr 1 1](https://cloud.githubusercontent.com/assets/921705/20632245/66d0dafc-b33c-11e6-867d-25eac8d0260d.jpeg)

**Fabric 1.7.0**
- Chrome Version 54.0.2840.98 (64-bit)
- enabledRetinaScaling: false
- devicePixelRatio: 2 (Retina display)
![fabric 1 7 0 dpr 2](https://cloud.githubusercontent.com/assets/921705/20632246/66d2f828-b33c-11e6-8054-40fc8a4bf124.jpeg)
![fabric 1 7 0 dpr 2 1](https://cloud.githubusercontent.com/assets/921705/20632244/66d0b8c4-b33c-11e6-9a6a-c501f0dc8c7e.jpeg)


## NON RETINA DISPLAY
**Fabric 1.5.0**
- Chrome Version 54.0.2840.98 (64-bit)
- enabledRetinaScaling: false (not supported in 1.5.0)
- devicePixelRatio: 1 (not supported in 1.5.0)
![fabric 1 5 0 dpr 1](https://cloud.githubusercontent.com/assets/921705/20632356/36d29a4c-b33d-11e6-857c-384c44181aac.jpeg)
![fabric 1 5 0 dpr 1 1](https://cloud.githubusercontent.com/assets/921705/20632357/395781c4-b33d-11e6-831b-b72b4aa842c6.jpeg)

**Fabric 1.7.0**
- Chrome Version 54.0.2840.98 (64-bit)
- enabledRetinaScaling: false
- devicePixelRatio: 2 (Retina display)
![fabric 1 7 0 dpr 1](https://cloud.githubusercontent.com/assets/921705/20632361/3d84f93e-b33d-11e6-8960-12a2000d3d79.jpeg)
![fabric 1 7 0 dpr 1 1](https://cloud.githubusercontent.com/assets/921705/20632363/42da2e22-b33d-11e6-9bbd-fd1bf2440c39.jpeg)

Feel free to test/compare your output:

Doodle:
http://benjaminmaertz.de/amcharts/fabricTest/index-1.5.0.html
http://benjaminmaertz.de/amcharts/fabricTest/index-1.7.0.html

Chart:
http://benjaminmaertz.de/amcharts/fabricTest/chart/index-1.5.0.html
http://benjaminmaertz.de/amcharts/fabricTest/chart/index-1.7.0.html


I would appreciate your help a lot – this is driving me crazy :)

**Comment by asturur** [comment]
This is a separate issue.
Would you mind trying to disable objectCaching for fabric 1.7.0
on your pathGroup use objectCaching = false and see if hte crispness is restored.

**Comment by asturur** [comment]
Or open a fiddle with public code so i can try it.

**Comment by maertz** [comment]
thanks for the quick response but it looks like there is no difference at all :/

			// SET OBJECTCACHING
			if ( location.href.indexOf("objectCaching") != -1 ) {
				g.objectCaching = false;
				g.set({
					objectCaching: false
				});
				console.log("disabled object caching");
			}

## RETINA (1.7.0)
The output seems to be the same...
![fabric 1 7 0 dpr 2 1](https://cloud.githubusercontent.com/assets/921705/20636795/17bac08a-b374-11e6-87fe-23be887718a1.jpeg)
![fabric 1 7 0 dpr 2](https://cloud.githubusercontent.com/assets/921705/20636796/17bb01a8-b374-11e6-97be-40c6ec54d521.jpeg)

## NON RETINA (1.7.0)
![fabric 1 7 0 dpr 1](https://cloud.githubusercontent.com/assets/921705/20636849/f825c2c8-b374-11e6-9d94-30cccfa20993.jpeg)
![fabric 1 7 0 dpr 1 1](https://cloud.githubusercontent.com/assets/921705/20636850/f825c34a-b374-11e6-8915-c74dd23be69e.jpeg)

## Test
http://benjaminmaertz.de/amcharts/fabricTest/index-1.7.0.html?objectCaching=true
http://benjaminmaertz.de/amcharts/fabricTest/chart/index-1.7.0.html?objectCaching=true

**Comment by asturur** [comment]
so, just to recap, the dataurl output is not the same as canvas on non retina display. i have a non retina screen so i think i can solve this, but i would love to get a jsfiddle with loading and export code

**Comment by asturur** [comment]
I cannot get any difference worth investigating with very simpe fiddle:
https://jsfiddle.net/jf03bbc3/3/

So please if you can produce a simple fiddle that recreate the blurrines open a separate issue.

1.7.0 canvas
![image](https://cloud.githubusercontent.com/assets/1194048/20639769/9d5c2bea-b3cc-11e6-955a-e6823ed02a6d.png)


1.7.0 dataurl

![image](https://cloud.githubusercontent.com/assets/1194048/20639761/769e0802-b3cc-11e6-8e13-5159f4772a37.png)

1.5.0 canvas
![image](https://cloud.githubusercontent.com/assets/1194048/20639774/c50d8724-b3cc-11e6-983a-f2a1c9a06e89.png)

1.5.0 dataurl
![image](https://cloud.githubusercontent.com/assets/1194048/20639775/cf89d432-b3cc-11e6-9016-f562a411fe4b.png)

**Comment by asturur** [comment]
i notice a very small blur on the bottom red line. That blurriness is some half pixel position and is not really a fabricjs fault but more a sort of trying to position elements in theyr original positions.

fabric 1.5.0 was parsing the svg as  width: 1100.5 height: 440.5 0 0 with a 1 border width
fabric 1.7.0 is parsing the svg as width 1100 height 670. that half pixel causes the blurrines.
If you position the object half pixel more left and down the blurriness disappear.

http://fabricjs.com/test/svg_import/  this is a page with many svg. if you observe it on a non retina screen do you notice blurriness?

**Comment by maertz** [comment]
Indeed, looks like the half pixel positioning is causing it. Thanks a lot for diving – I appreciate your help!

I wrote a little helper to cancel these half pixel thingies and the output is much better now.

http://jsfiddle.net/1f0ubpn8/1/

<img width="322" alt="screen shot 2016-11-28 at 13 15 42" src="https://cloud.githubusercontent.com/assets/921705/20668015/cfebbc28-b56c-11e6-87fd-38f6c2d01c29.png">
<img width="321" alt="screen shot 2016-11-28 at 13 15 49" src="https://cloud.githubusercontent.com/assets/921705/20668016/cfebf8aa-b56c-11e6-8e2d-0ea24c4e726f.png">

**Comment by asturur** [comment]
i have to find out where it comes from and if it is right.
Not really sure what i should do.

2016-11-28 13:16 GMT+01:00 Bny <notifications@github.com>:

**Comment by maertz** [comment]
Looks like the coordinates system of the canvas works a little bit different than "expected".

Through a quick look up I found this interesting article:
http://diveintohtml5.info/canvas.html#pixel-madness

**Comment by MarcoMiltenburg** [comment]
What's the status of this?  Does this now mean we have to start aligning/rounding to half pixels to prevent anti-aliasing?  This sounds like a pretty fundamental problem that should be handled by the library (or even should not happen at all while drawing the canvas if you ask me).

**Comment by maertz** [comment]
I've fix this issue temporary by shifting all things by 0.5 pixel - worked well but comfortable is different.

**Comment by asturur** [comment]
is not that easy.
when rotation and scaling come in, half pixel is not an options. when you
use centering position you need half pixel just if the width is odd or even.

when canvas is zoomed and tralsated you do not know on what pixel will you
render.

sometimes you work on a big canvas in a small space and you have 10 pixels
in the space of one pixel. you cannot align everything to avoid aliasing

**Comment by MarcoMiltenburg** [comment]
I recently upgraded a project from 1.4 to 1.7 and it was not a problem with 1.4. So something must have changed in 1.5 or up that causes this.

**Comment by asturur** [comment]
yes, the strokewidth is included in the object bounding box in 1.5
This causes a shift of 0.5 pixel with the default strokewidth of 1.
But fixes lot of other things that fabric does.
Other than that from 1.4.11 upward, zoom and panning started to work better, people started to use it more and those defects are more visible.

staying on 1.4 is not an option from my point of view.
You can set strokeWidth to 0 in the prototype and disable caching if you want to have a 1.4 experience like but have the newest fixes and functions.

I have to investigate if ctxImageSmoothScaling set to false can fix the blurring problems at expense of general uglyness.

**Comment by MarcoMiltenburg** [comment]
I've had some more time to dig in but even the simplest of code (without a border) causes blurry lines. Look at this JSFiddle: http://jsfiddle.net/dosmLq0n/

Even this causes blurry lines on non-retina screens:

![image](https://cloud.githubusercontent.com/assets/6213937/23510372/a156ddfc-ff59-11e6-8697-464d4726d1f5.png)

When I open it on my retina MBP it renders sharp. But shouldn't this render just fine on non-retina screens as well ??

**Comment by MarcoMiltenburg** [comment]
@maertz That trick seems to work on Windows 10 but not on Windows 7 or macOS 10.11 on a non-retina MBP or even with macOS 10.12 on a retina MBP in non-retina mode (i.e. with an external monitor attached), all using the same Chrome 56 browser.

**Comment by asturur** [comment]
http://jsfiddle.net/dosmLq0n/1/

here there are 2 examples of things rendering good.

as you can see, any other number combination will render bad.
There is no much thing i can do.
I can maybe improve the positioning over the cachecanvas ( maybe ) to achieve the same number of blurried/unblurried cases as with cache disabled.

but blur will be there for half pixels

**Comment by MarcoMiltenburg** [comment]
Thanks, that seems to fix the issue. Object caching is not really important for this project. 

The blurring problem only seems to happen with objects that are drawn, e.g. rects, circles or paths but not e.g. with text or SVG's that are loaded dynamically. Is this correct?  Just checking, otherwise I have a lot more code to check / fix.

**Comment by asturur** [comment]
blurrying happens anyway. in some shape is not evident.
a circle for example is either pixellated or blurry.

that bluriness is anti aliasing, when anti aliasing happens twice, looks bad

**Comment by heaxo** [comment]
@asturur 

https://github.com/fabricjs/fabric.js/assets/72446584/1ca5b15a-4b81-44c8-a0f3-98f44e39491d

Hello, I would like to know how I should handle this situation. I know it shouldn't be a problem with the fabric, but I haven't found a solution. Can you help me? The general problem is that when I zoom in the browser to 25%, I use fabricjs to render the text, but when I zoom in to 500%, it becomes blurry. When I directly refresh the browser and re render the fabric at 500%, it becomes very clear. What can I do to keep it clear all the time?

Here is my code
`Const canvas=new fabric Canvas ('canvas');
Fabric Object. prototype. objectCaching=false;
Canvas. loadFromJSON (JSON, canvas. renderAll. bind (canvas));`

**Comment by asturur** [comment]
@heaxo on top of my head i don't remember if there is some event that fires when you change the browser zoom.
If you find such event on the browser i can help you set up some code to keep the canvas sharp

**Comment by heaxo** [comment]
@asturur Hello, I need to zoom in and out events in the browser to keep the fabric clear. Can you help me with this while listening to events
Window. addEventListener ('resize ', ()=>{
});
Thank you very much for your help
